### ISSUE TITLE
vnode.domSize is wrong in fragments when a child has a delayed removal

### ISSUE THREAD

**Issue by pygy** [MithrilJS/mithril.js#2780]
**Mithril.js version:**
v1.0 -> v2.2

## Code

```javascript
const a = m.fragment({}, [
  'a',
  m('', {
    onbeforeremove({dom, domSize}){
      return new Promise(()=>{})
    }
  }, 'b'),
  'c'
])

const b = m.fragment({onupdate({dom, domSize}){
  console.log("parent", {dom, domSize})
  console.log(document.body.childNodes.length)
}}, [
  'a',
  null,
  'c'
])

m.render(document.body, a)
m.render(document.body, b)
```

[flems](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgCsEAaEAYwHs0AXGWvAHTSrThoAIMOBeDrfGABOGAOY5aACmABfEh2TMOHAOQYVJJf0kqNHYFuXUARjDCUhMS1koA3GNIAmlLPOdYAyhABeMGQEoDNGUQjksaAFchYLQYAHcOAAUhFwg4B0l-HgA+WX9DDhktOVVjFX9NYNVyFWYAXXy0ZlZ2DmNefkERcXoaaWoIgAdHDDonFzcXL18AoOUWylh8KEpRSUYQQYxLWg35YHdJzx8-RvnqOEWYZdXJZ3IIiRp8Y0pHAE98cgALaEcAOTe8GW9FENG+jRkJUUVTUGi0aAiUCglWUKhq9UazAEO0cVjulAeTxeb3e8gwjRx9DxQgJRN6JI+8mMjRAZHSsHINAgFzwACY+YgAAwAWgAjMKQDI6mQoBA0ABrBAodDYXCIEBYCDgoTQNkgKJQPDfGg0QZwRAAektETQgwVoi+LktWp10AAAnz8F6+S7td9dVB8Fq0ERSCAaO9BuqQHByLrBjQpXUZEA)

## Steps to Reproduce
1. render a fragment with several nodes (`[a, b, c]`), one of them (`b`) with an `onbeforeremove` hook that delays node removal.
2. render the same fragment, but with said node removed from vdom check the fragment's `vnode.domSize` in `onupdate`.

## Context
We should provide users a way to reach the nodes of the fragment. Both the nodes in vdom and the nodes present in DOM. Currently the docs recommends to rely on `domSize` for fragments, but doing so would let users grab `[a, b]`, not `[a, c]` nor `[a,b,c]` which would both be less wrong.

Both `[a, c]` and `[a,b,c]` could be retrieved by scanning vdom. Providing an iterator would seem ideal to me.

Someting like this (untested) helper would let folks iterate using `for (dom of domFor(vnode))` loops.

```js
const delayedRemoval = new WeakSet

function* domFor({dom, domSize}) {
  if (domSize == null) {
    if !delayedRemoval.has(dom) yield dom
  } else {
    let i = 0
    while (i < domSize) {
      if (!delayedRemoval.has(dom)) {
        yield dom
        i++
      }
      dom = dom.nextSibling
    }
  }
}
```
At the very least, the limitation should be documented.

**Comment by pygy** [comment]
Note that `domFor` could be made tremendously simpler by adding the nodes with delayed removal to a WeakMap, and use plain `nexSibling` logic while skipping the now identifiable nodes.

**Comment by dead-claudia** [comment]
> At the very least, the limitation should be documented.

Yeah, it should be documented. Actually fixing `vnode.domSize` to work in the face of partial component removal would require us to do some serious changes not just to the removal logic and (to a lesser extent) patching logic, but also add at minimum a field to the vnode structure itself. And yes, we'd need to also update every intermediate fragment up to and including the parent element after removal.

Or, long story short, not trivial at all.

**Comment by dead-claudia** [comment]
> * For both, we'd have to `delayedRemoval.delete(vnode.dom)` on every insertion, or cached vnodes could behave problematically. It may degrade perf. We just could state that cached DOM nodes can't be removed asynchronously.

That is an option - https://mithril.js.org/vnodes.html#avoid-memoizing-mutable-vnodes already declares one other instance where behavior is undefined. (You wouldn't see anything weird until you attempt to remove it or return a different vnode, though.)

> AFAICT this would work, only with local info and the `toBeRemoved` map.
> 
> The docs currently says that a node with delayed removal is only removed when the promise returns. There's no mention of parent fragment or component removal.
> 
> Mithril behaves accordingly. Suppose `b` has delayed removal. If you `render( x, [a, b, c])` then `render(x, [a, null, c])` then `render(x, null)`, `b` will stay in tree until its promise resolves.

Yeah, we should document and clarify the intended behavior that occurs with async removal:

- Parent vnodes' removals are *not* blocked. Only the delayed vnode and its children are delayed.
- Parent fragments still have their `onremove` callbacks fired and are removed from the tree.
- Children of delayed vnodes do *not* have `onbeforeremove` callbacks invoked. They are simply removed (with `onremove` callbacks called) immediately upon their respective delayed vnode's `onbeforeremove` promise resolving.
- Delayed vnodes exist in a state where they are detached from any parent vnode yet their underlying DOM nodes remain attached to the underlying parent DOM node.

We should also verify the above behavior, specifically that children of delayed parents have their removal hooks delayed until after the delayed removal promise resolves. (That should've been the previous behavior and still is the behavior we want IMHO, but I *don't* think that's what's implemented in #2782.)
